### PR TITLE
Better handle pkg updates with name and namespace changes

### PR DESCRIPTION
**Pull request by phanimarupaka**
This PR does the following:

1. Ensures line comment with format `kpt-merge: namespace/name` on `metadata` key node of each resource after the package is fetched using `kpt pkg get` as PostRunE
2. Update `merge3` logic to merge resources based on GKNN+filepath where NN is derived from the merge comment.

This should be merged after https://github.com/kubernetes-sigs/kustomize/pull/3821 and updating the kyaml dependency. Till then tests will fail.

### CODE REVIEW

**Comment by mortent** [comment]
I'm not convinced we want to fork the 3-way merge library into `third_party`. We have done it for the cli commands because we want to control the cli experience in kpt, but this is different. It is also hard to see from the PR what is actually taken from the library and what are the changes since it is all in one commit. Isn't this something we could contribute upstream?
Can you split this into two separate commits so it is easier to see what is going on? And when we merge it, we should first do a separate commit which is a clean copy and then add a second commit with our changes. It makes it easier to see in the future how we have diverged from the source. This is a situation where the merge commits is a bit unfortunate.. 

It would also be useful to include a description of the behavior, essentially just summarize the results of the test cases.

**Comment by phanimarupaka** [comment]
> I'm not convinced we want to fork the 3-way merge library into `third_party`. We have done it for the cli commands because we want to control the cli experience in kpt, but this is different. It is also hard to see from the PR what is actually taken from the library and what are the changes since it is all in one commit. Isn't this something we could contribute upstream?

Making changes to kyaml/merge3 would break current package workflows of users as it will be pointed by current version of kpt. So we should fork that high level merge3 logic to Kpt next. Entire merge logic is not moved to next.

> Can you split this into two separate commits so it is easier to see what is going on? And when we merge it, we should first do a separate commit which is a clean copy and then add a second commit with our changes. It makes it easier to see in the future how we have diverged from the source. This is a situation where the merge commits is a bit unfortunate..
> 
Done.
> It would also be useful to include a description of the behavior, essentially just summarize the results of the test cases.
The only rule is that there should not be 2 resources with same GVK in same file. That is an error case. Rest cases should work the same as per current expectations. Since resources are not identified by `name` and `namespace` anymore, functions like `set-namespace`, `set-nameprefix` would not affect merging of resources.

**Comment by mortent** [comment]
> Making changes to kyaml/merge3 would break current package workflows of users as it will be pointed by current version of kpt. So we should fork that high level merge3 logic to Kpt next. Entire merge logic is not moved to next.

Yeah, I agree we can't change the behavior in upstream. I think the question is whether whether we could make this an extension point in kyaml. This would allow us to use a proper way to extend the behavior. With this solution we are essentially carrying a patch, which could be a bit painful long term and doesn't really scale well.

> The only rule is that there should not be 2 resources with same GVK in same file which is an error case. Rest cases should work the same as per current expectations. Since resources are not identified by `name` and `namespace` anymore, functions like `set-namespace`, `set-nameprefix` would not affect merging of resources. If any of GVK is changed, it is treated as new resource and both such resources end up on local.

We probably should only use GK for this, updating the version of the resource seems more likely than changing the Group or Kind.
I think the key change this brings is that for all kpt packages, we now have the restriction that a file can not include multiple resources of the same GK. This doesn't seem to unreasonable for applications, although having multiple services for a single application is somewhat common I think. I'm more concerned about KCC resources, where this restriction might be a bit more inconvenient. I guess including the index could address this, but it would bring with it a whole set of other issues..

I think we want input from others too on this: @droot @Shell32-Natsu

**Comment by phanimarupaka** [comment]
> > Making changes to kyaml/merge3 would break current package workflows of users as it will be pointed by current version of kpt. So we should fork that high level merge3 logic to Kpt next. Entire merge logic is not moved to next.
> 
> Yeah, I agree we can't change the behavior in upstream. I think the question is whether whether we could make this an extension point in kyaml. This would allow us to use a proper way to extend the behavior. With this solution we are essentially carrying a patch, which could be a bit painful long term and doesn't really scale well.
> 
Agree. I can make this an extension in kyaml which is not a big issue.

> > The only rule is that there should not be 2 resources with same GVK in same file which is an error case. Rest cases should work the same as per current expectations. Since resources are not identified by `name` and `namespace` anymore, functions like `set-namespace`, `set-nameprefix` would not affect merging of resources. If any of GVK is changed, it is treated as new resource and both such resources end up on local.
> 
> We probably should only use GK for this, updating the version of the resource seems more likely than changing the Group or Kind.
> I think the key change this brings is that for all kpt packages, we now have the restriction that a file can not include multiple resources of the same GK. This doesn't seem to unreasonable for applications, although having multiple services for a single application is somewhat common I think. I'm more concerned about KCC resources, where this restriction might be a bit more inconvenient. I guess including the index could address this, but it would bring with it a whole set of other issues..
> 
> I think we want input from others too on this: @droot @Shell32-Natsu

You are right. Adding index to the list leads to merging totally unrelated resources in some cases.

**Comment by droot** [comment]
Two big decisions we are making in this PR:

- Forking merge3 logic from kyaml to kpt repo
- Introducing new constraint that a file should not have two resources with the same GVK ( for ex. file can not have two deployments)

## Forking merge3 logic from kyaml to kpt repo

Our current understanding is that we should try to keep the lower level building blocks in kyaml. Forking is the last resort when we have exhausted all possible ways of introducing new changes in the upstream. So if new merge logic can be supported by extending the upstream library, great, we should do that.

## Introducing new constraint

I haven't formed any strong opinion about it yet other than that it takes away flexibility of organizing the resources the way users want (btw, that's a big cost to pay, so we need to ensure benefits outweighs the cost) We need to think through how it impacts all config management workflows not just the `pkg update` workflow. I am assuming users will only come to know about this constraint when they do `pkg update`. Ideally, they should come to know about it much earlier (think of `kpt vet or lint` or even `kpt get` warning about that the pkg violates some constraint). TL;DR is we need to understand the impact of this constraint better and how can we do that ?

File: internal/cmdget/cmdget.go
Comment (mortent) [inline internal/cmdget/cmdget.go:108]: This seems like a pretty important part of kpt functionality, so I don't think we should put it as a step in the cobra command. We should do this as part of the `get.Command` functionality.
Also, this must also happen for update. We might be pulling in new resources from upstream that doesn't have the comment.

File: internal/cmdget/cmdget_test.go
Comment (mortent) [inline internal/cmdget/cmdget_test.go:404]: Why are we adding this test here? it seems like it only tests the code in the `addmergecomment` package.

File: internal/util/addmergecomment/addmergecomment.go
Comment (mortent) [inline internal/util/addmergecomment/addmergecomment.go:44]: I don't think this is correct. Cluster-scoped resources shouldn't have a namespace and an empty namespace value for namespaced resource doesn't always resolve to "default".

File: internal/util/addmergecomment/addmergecomment.go
Comment (mortent) [inline internal/util/addmergecomment/addmergecomment.go:46]: Should we also allow putting the comment as a HeadComment?

File: internal/util/merge/merge3.go
Comment (mortent) [inline internal/util/merge/merge3.go:221]: I think there are libraries in kubernetes that does this. We should consider just using them.

File: internal/util/merge/merge3.go
Comment (mortent) [inline internal/util/merge/merge3.go:235]: Like mentioned above, I don't think we should assume a default namespace if the metadata.Namespace field is not set.

File: internal/util/addmergecomment/addmergecomment.go
Comment (phanimarupaka) [reply internal/util/addmergecomment/addmergecomment.go:48]: I don't think so. That leads to confusion. It is consistent with setter comments.

File: internal/util/merge/merge3.go
Comment (phanimarupaka) [reply internal/util/merge/merge3.go:221]: I din't find any. This is simple enough IMO.

File: internal/util/merge/merge3.go
Comment (phanimarupaka) [reply internal/util/merge/merge3.go]: updated.

File: internal/util/addmergecomment/addmergecomment.go
Comment (phanimarupaka) [reply internal/util/addmergecomment/addmergecomment.go]: Updated.

File: internal/cmdget/cmdget_test.go
Comment (phanimarupaka) [reply internal/cmdget/cmdget_test.go]: Moved to testutil

File: internal/cmdget/cmdget.go
Comment (phanimarupaka) [reply internal/cmdget/cmdget.go]: Updated.

File: thirdparty/kyaml/merge3/merge3_test.go
Comment (phanimarupaka) [reply thirdparty/kyaml/merge3/merge3_test.go]: Yes.

File: internal/util/update/resource-merge.go
Comment (mortent) [inline internal/util/update/resource-merge.go:130]: I don't follow why we are adding the merge comments at these locations in the code. It seems like we should decide on a strategy for when we should add these comments, and then based on that decide where in the codebase we should make the call.

So I know there are at least two cases:
 * If we pull in new resources in the local fork that doesn't already have the merge comment, we want to add it.
 * Sometimes we want to decide if there are local changes to a package and we don't want any merge comments added in the local fork to be interpreted as local changes to a fork that would prevent deletion. 
I'm not sure this is complete.

Based on this, it seems like we should add any necessary merge comments after a package has been updated/fetched, regardless of how it happened (merge strategy, new subpackage added...).
For the diff situation, it seems like we should either make these changes consistently for every package during updates, or handle it within the diff functionality. This seems like we update the packages with this information in just some situations.

Comment (phanimarupaka) [reply internal/util/update/resource-merge.go]: Agree. These should be added in the update.go. Before we invoke merge operation, we should get all the three sources on level ground by adding merge comments to all of them so that the update would work just fine like today. 

So in update workflow these are invoked twice:
1. After doing pre-checks and just before invoking updaters of individual strategies.
2. At the end, to make sure that the all updated resources on local have the merge comments as a final ensuring step.

Individual strategies should not have to worry about this.

File: internal/util/update/update.go
Comment (phanimarupaka) [reply internal/util/update/update.go]: Moved it towards the end of Run method.

File: internal/util/update/resource-merge.go
Comment (phanimarupaka) [reply internal/util/update/resource-merge.go]: removed.

File: internal/util/merge/merge3.go
Comment (phanimarupaka) [reply internal/util/merge/merge3.go:222]: Kubernetes dependencies are painful. Used kyaml library instead. It has a util method too.

File: internal/util/update/update.go
Comment (phanimarupaka) [reply internal/util/update/update.go]: I want to add it to a place where all the packages are fetched ensured that they are present. Moved it as first step to method.

File: internal/testutil/testutil.go
Comment (phanimarupaka) [reply internal/testutil/testutil.go]: True. I made it configurable and added addMergeCommentToSource as option to the Diff method. Not adding comments to dest as it is expected to have comments. So this would test that the comments are added for cases where comments should be added.

Callers like get_test, update_test would pass true for addMergeCommentToSource as comments should be added after these operations
Callers like fetch_test, resource-merge_test don't are elemental and don't add merge comments. So addMergeCommentToSource is false for them when repos are compared.

File: internal/testutil/testutil.go
Comment (mortent) [inline internal/testutil/testutil.go:105]: I find this name confusing. What this has to do with `expected` is unclear.

File: internal/util/diff/pkgdiff.go
Comment (mortent) [inline internal/util/diff/pkgdiff.go:139]: This function seems identical to the function in `testutil`. Also, the name is confusing.

File: internal/util/diff/pkgdiff.go
Comment (mortent) [inline internal/util/diff/pkgdiff.go:144]: Do we absolutely need to do a copy here? I have noticed that file copying actually matters for how long it takes for kpt to get/update a package, so we should try to minimize if possible.

File: internal/util/update/update.go
Comment (mortent) [reply internal/util/update/update.go:422]: Can we remove it here then? It seems like doing at the start of the function should be sufficient?

File: internal/testutil/testutil.go
Comment (phanimarupaka) [reply internal/testutil/testutil.go]: My bad. Moved it to addmergecomments package itself. It is cleaner now.

File: internal/util/diff/pkgdiff.go
Comment (phanimarupaka) [reply internal/util/diff/pkgdiff.go]: Good catch. We don't need it. Removed it.

File: internal/util/diff/pkgdiff.go
Comment (phanimarupaka) [reply internal/util/diff/pkgdiff.go]: Yes. Moved it to addmergecomments package itself. It is cleaner now.

File: internal/util/update/update.go
Comment (phanimarupaka) [reply internal/util/update/update.go]: Sorry. Missed removing it. Done.

**Review by mortent — APPROVED** [review]

**Pull request merged**
